### PR TITLE
Add student scores + work (component grading) to milestone reports

### LINKED ISSUE #2150: Add student scores to milestone reports
As of now, we only show student completion in the milestone reports. We should show the grading view (response, status + score/feedback fields) for each student in place of progress.

Target completion: Jan 2020

### PR DESCRIPTION
**Pull request by breity**
Milestone reports (once the milestone's completion threshold has been met) now show two tabs below the report content: Recommendations (if authored) and Student Work. Student Work shows the node grading view for the component that the milestone addresses.

To test:
1. Run ```npm install``` to update dependencies.
1. Set up a run with a milestone report. (Here is a sample unit with a milestone report enabled: 
[212.zip](https://github.com/WISE-Community/WISE/files/4036817/212.zip)).
1. Add two or more students to the run and complete the milestone component.
1. Open the teacher tools for the run and verify that once the completion thresholds have been met, the report content and summary graphs appear for the milestone and that the Recommendations and Student Work tabs show up. Verify that the Student Work display shows c-rater score and work for each workgroup correctly.
1. From the teacher's milestone display, make sure sending annotations works. (Note that changing the score for a workgroup will remove that workgroup's data from the milestone report summary and graphs. This is because milestone reports only look at the latest student score and include it if the type is autoScore. I'm creating a separate issue to address this.)

Closes #2150.

### CODE REVIEW

File: src/main/webapp/wise5/authoringTool/project/projectController.js [hunk 1]
```diff
@@ -160,20 +162,17 @@ class ProjectController {
...
-        } else if (stateName == 'root.project.node') {
-        } else if (stateName == 'root.project.asset') {
-          this.saveEvent('assetsViewOpened', 'Navigation');
-        } else if (stateName == 'root.project.info') {
-          this.saveEvent('projectInfoViewOpened', 'Navigation');
-        } else if (stateName == 'root.project.notebook') {
-          this.saveEvent('notebookViewOpened', 'Navigation');
-        }
+    this.$transitions.onSuccess({}, ($transition) => {
+      let stateName = $transition.name;
```
Comment (hirokiterashima) [inline src/main/webapp/wise5/authoringTool/project/projectController.js]:
```suggestion
      const stateName = $transition.name;
```

File: src/main/webapp/wise5/authoringTool/project/projectController.js [hunk 2]
```diff
@@ -160,20 +162,17 @@ class ProjectController {
...
+      if (stateName == 'root.project') {
```
Comment (hirokiterashima) [inline src/main/webapp/wise5/authoringTool/project/projectController.js]:
```suggestion
      if (stateName === 'root.project') {
```

File: src/main/webapp/wise5/authoringTool/project/projectController.js [hunk 3]
```diff
@@ -160,20 +162,17 @@ class ProjectController {
...
+        this.saveEvent('projectHomeViewOpened', 'Navigation');
+      } else if (stateName == 'root.project.node') {
```
Comment (hirokiterashima) [inline src/main/webapp/wise5/authoringTool/project/projectController.js]:
```suggestion
```

File: src/main/webapp/wise5/authoringTool/project/projectController.js [hunk 4]
```diff
@@ -160,20 +162,17 @@ class ProjectController {
...
+      } else if (stateName == 'root.project.asset') {
```
Comment (hirokiterashima) [inline src/main/webapp/wise5/authoringTool/project/projectController.js]:
```suggestion
      } else if (stateName === 'root.project.asset') {
```

File: src/main/webapp/wise5/authoringTool/project/projectController.js [hunk 5]
```diff
@@ -160,20 +162,17 @@ class ProjectController {
...
+        this.saveEvent('assetsViewOpened', 'Navigation');
+      } else if (stateName == 'root.project.info') {
```
Comment (hirokiterashima) [inline src/main/webapp/wise5/authoringTool/project/projectController.js]:
```suggestion
      } else if (stateName === 'root.project.info') {
```

File: src/main/webapp/wise5/authoringTool/project/projectController.js [hunk 6]
```diff
@@ -160,20 +162,17 @@ class ProjectController {
...
+        this.saveEvent('projectInfoViewOpened', 'Navigation');
+      } else if (stateName == 'root.project.notebook') {
```
Comment (hirokiterashima) [inline src/main/webapp/wise5/authoringTool/project/projectController.js]:
```suggestion
      } else if (stateName === 'root.project.notebook') {
```

File: src/main/webapp/wise5/classroomMonitor/classroomMonitorComponents/nodeGrading/nodeGradingView/nodeGradingView.js [hunk 7]
```diff
@@ -17,106 +16,111 @@ class NodeGradingController {
...
+                const workgroupId = studentWork.workgroupId;
+                const nodeId = studentWork.nodeId;
                 if (nodeId === this.nodeId && this.workgroupsById[workgroupId]) {
-                    // a workgroup has a new componentState for this node
                     this.updateWorkgroup(workgroupId);
                 }
             }
         });
 
         // save event when node grading view is displayed and save the nodeId that is displayed
```
Comment (hirokiterashima) [inline src/main/webapp/wise5/classroomMonitor/classroomMonitorComponents/nodeGrading/nodeGradingView/nodeGradingView.js]:
```suggestion
```

File: src/main/webapp/wise5/classroomMonitor/classroomMonitorComponents/nodeGrading/nodeGradingView/nodeGradingView.js [hunk 8]
```diff
@@ -17,106 +16,111 @@ class NodeGradingController {
...
-        let context = "ClassroomMonitor", nodeId = this.nodeId, componentId = null, componentType = null,
+        const context = "ClassroomMonitor", nodeId = this.nodeId, componentId = null, componentType = null,
```
Comment (hirokiterashima) [inline src/main/webapp/wise5/classroomMonitor/classroomMonitorComponents/nodeGrading/nodeGradingView/nodeGradingView.js]: Use single quotes

File: src/main/webapp/wise5/classroomMonitor/classroomMonitorComponents/nodeGrading/nodeGradingView/nodeGradingView.js [hunk 9]
```diff
@@ -17,106 +16,111 @@ class NodeGradingController {
...
+            } else {
+                return 0;
+            }
+        } else {
+            return this.ProjectService.getMaxScoreForNode(this.nodeId);
+        }
+    }
+
+    getHiddenComponents() {
+        let hiddenComponents = [];
```
Comment (hirokiterashima) [inline src/main/webapp/wise5/classroomMonitor/classroomMonitorComponents/nodeGrading/nodeGradingView/nodeGradingView.js]:
```suggestion
        const hiddenComponents = [];
```

File: src/main/webapp/wise5/classroomMonitor/classroomMonitorComponents/nodeGrading/nodeGradingView/nodeGradingView.js [hunk 10]
```diff
@@ -17,106 +16,111 @@ class NodeGradingController {
...
+        if (this.nodeContent) {
```
Comment (hirokiterashima) [inline src/main/webapp/wise5/classroomMonitor/classroomMonitorComponents/nodeGrading/nodeGradingView/nodeGradingView.js]: Is this check necessary?

File: src/main/webapp/wise5/classroomMonitor/classroomMonitorController.js [hunk 11]
```diff
@@ -144,7 +146,7 @@ class ClassroomMonitorController {
         });
 
         // listen for state change events
```
Comment (hirokiterashima) [inline src/main/webapp/wise5/classroomMonitor/classroomMonitorController.js]:
```suggestion
```

File: src/main/webapp/wise5/classroomMonitor/classroomMonitorController.js [hunk 12]
```diff
@@ -144,7 +146,7 @@ class ClassroomMonitorController {
...
-        this.$rootScope.$on('$stateChangeSuccess', (event, toState, toParams, fromState, fromParams) => {
+        $transitions.onSuccess({}, ($transition) => {
             // close the menu when the state changes
```
Comment (hirokiterashima) [inline src/main/webapp/wise5/classroomMonitor/classroomMonitorController.js]:
```suggestion
```

**Comment by hirokiterashima** [comment]
Also found a new issue where the AT would not load. There's an error in the js console:
```
angular.js:15567 TypeError: Cannot read property '$broadcast' of undefined
    at AuthoringToolController.turnOffJSONValidMessage (authoringToolController.js:268)
    at AuthoringToolController.processUI (authoringToolController.js:264)
    at new AuthoringToolController (authoringToolController.js:109)
    ...
```

**Comment by breity** [comment]
@hirokiterashima Thanks, I implemented the code changes and fixed the AT loading bug.

**Comment by breity** [comment]
@geoffreykwan @hirokiterashima Okay, milestones always show up now even if teacher changes the score for all the students. Milestone report data now also takes into account teacher scores.

File: src/main/webapp/wise5/classroomMonitor/milestones/milestonesController.js [hunk 13]
```diff
@@ -68,6 +68,21 @@ class MilestonesController {
...
       }
     });
+
+    this.$scope.$on('annotationReceived', (event, args) => {
+      const annotation = args.annotation;
+      if (annotation) {
+          const nodeId = annotation.nodeId;
+          const componentId = annotation.componentId;
+          for (const projectAchievement of this.projectAchievements) {
+            if (projectAchievement.nodeId && projectAchievement.componentId &&
```
Comment (geoffreykwan) [inline src/main/webapp/wise5/classroomMonitor/milestones/milestonesController.js]: Is the check for ```projectAchievement.nodeId && projectAchievement.componentId``` necessary if we're already checking for
```
projectAchievement.nodeId === nodeId && projectAchievement.componentId === componentId
```

File: src/main/webapp/wise5/classroomMonitor/milestones/milestonesController.js [hunk 14]
```diff
@@ -472,14 +487,32 @@ class MilestonesController {
...
+          this.addDataToAggregate(aggregate, mergedAnnotation);
+        }
       }
     }
     return aggregate;
   }
 
+  mergeAutoScoreAndTeacherScore(autoScoreAnnotation, teacherScoreAnnotation) {
+    if (autoScoreAnnotation.data.scores) {
+      for (let subScore of autoScoreAnnotation.data.scores) {
```
Comment (geoffreykwan) [inline src/main/webapp/wise5/classroomMonitor/milestones/milestonesController.js]: I think we can change let to const here.

**Comment by breity** [comment]
@geoffreykwan Made requested code cleanup changes.

File: src/main/webapp/wise5/classroomMonitor/milestones/milestonesController.js [hunk 15]
```diff
@@ -68,6 +68,20 @@ class MilestonesController {
...
+          const nodeId = annotation.nodeId;
+          const componentId = annotation.componentId;
+          for (const projectAchievement of this.projectAchievements) {
+            if (projectAchievement.nodeId === nodeId && 
+                projectAchievement.componentId === componentId) {
+              this.updateMilestoneCompletion(projectAchievement.id);
+            }
+          }
+      }
+  });
```
Comment (geoffreykwan) [inline src/main/webapp/wise5/classroomMonitor/milestones/milestonesController.js]: Would it be ok to change this block to
```
    this.$scope.$on('annotationReceived', (event, args) => {
      for (const projectAchievement of this.projectAchievements) {
        if (projectAchievement.nodeId === args.annotation.nodeId && 
            projectAchievement.componentId === args.annotation.componentId) {
          this.updateMilestoneCompletion(projectAchievement.id);
        }
      }
    });
```

Comment (breity) [reply src/main/webapp/wise5/classroomMonitor/milestones/milestonesController.js]: Ah yeah, probably. We'll never have a case where `args.annotation` is undefined, right?

Comment (geoffreykwan) [reply src/main/webapp/wise5/classroomMonitor/milestones/milestonesController.js]: If the event is 'annotationReceived' theoretically it should always send it with the annotation.

Comment (breity) [reply src/main/webapp/wise5/classroomMonitor/milestones/milestonesController.js]: Cool, just pushed this.

**Pull request merged**